From d8b3a7dea639913eae3fdacc0b9a4dce1c9e7ed2 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Date: Tue, 12 Jan 2021 14:42:46 +0200
Subject: [PATCH 194/269] plat: s32g: Place BL33 in DDR

This moves U-Boot from SRAM to DDR right before BL31 and reserve 5MB for
it.

Issue: ALB-6294
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32g/include/platform_def.h | 22 +++++++++++++---------
 plat/nxp/s32g/s32g_bl31.c            |  8 ++++++--
 2 files changed, 19 insertions(+), 11 deletions(-)

diff --git a/plat/nxp/s32g/include/platform_def.h b/plat/nxp/s32g/include/platform_def.h
index b80b15f47..42e137cd4 100644
--- a/plat/nxp/s32g/include/platform_def.h
+++ b/plat/nxp/s32g/include/platform_def.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2020 NXP
+ * Copyright 2019-2021 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -15,7 +15,7 @@
 #include <common_def.h>
 #include <tbbr_img_def.h>
 
-#define SIZE_1M		(1024 * 1024)
+#define SIZE_1M		(0x100000)
 
 /* MPIDR_EL1 for the four A53 cores is as follows:
  *	A53_0_cpu0:	0x8000_0000
@@ -125,7 +125,7 @@
  * the Linux device-tree.
  */
 #define S32G_PMEM_END		(0xffffffff)
-#define S32G_PMEM_LEN		0x00200000	/* conservatively allow 2MB */
+#define S32G_PMEM_LEN		(2 * SIZE_1M)	/* conservatively allow 2MB */
 #define S32G_PMEM_START		(S32G_PMEM_END - S32G_PMEM_LEN + 1)
 
 /* Physical address 0x0 is actually mapped; to increase our
@@ -143,7 +143,8 @@
 
 #define BL31SRAM_BASE		BL2_BASE
 #define BL31SRAM_MAX_PAGES	50
-#define BL31SRAM_LIMIT		(BL31SRAM_BASE + BL31SRAM_MAX_PAGES * PAGE_SIZE)
+#define BL31SRAM_SIZE		(BL31SRAM_MAX_PAGES * PAGE_SIZE)
+#define BL31SRAM_LIMIT		(BL31SRAM_BASE + BL31SRAM_SIZE)
 
 #define BL31SSRAM_IVT		(S32G_SSRAM_BASE)
 #define BL31SSRAM_IVT_SIZE	(0x140)
@@ -159,11 +160,14 @@
 /* U-boot addresses in SRAM. BL33_DTB and BL33_ENTRYPOINT must be kept in
  * sync with u-boot's CONFIG_DTB_SRAM_ADDR and CONFIG_SYS_TEXT_BASE.
  */
-#define BL33_DTB		(S32G_SRAM_BASE + 0x90000)
-#define BL33_ENTRYPOINT		(S32G_SRAM_BASE + 0xa0000)
+#define S32G_BL33_IMAGE_SIZE	(5 * SIZE_1M)
+/* Leave a gap between BL33 and BL31 to avoid MMU entries merge */
+#define BL33_BASE		(S32G_PMEM_START - S32G_BL33_IMAGE_SIZE - \
+				 SIZE_1M)
+#define BL33_DTB		(BL33_BASE + 0x90000)
+#define BL33_ENTRYPOINT		(BL33_BASE + 0xa0000)
 #define S32G_BL33_IMAGE_BASE	(BL33_DTB)
-#define S32G_BL33_LIMIT		(S32G_SRAM_END)
-#define S32G_BL33_IMAGE_SIZE	(S32G_BL33_LIMIT - S32G_BL33_IMAGE_BASE)
+#define S32G_BL33_LIMIT		(BL33_BASE + S32G_BL33_IMAGE_SIZE)
 
 /* BL31 location in DDR - physical addresses only, as the MMU is not
  * configured at that point yet
@@ -204,7 +208,7 @@
 #if defined IMAGE_BL31
 #define FIRMWARE_WELCOME_STR_S32G_BL31	"This is S32G BL31\n"
 /* To limit usage, keep these in sync with sizeof(s32g_mmap) */
-#define MAX_MMAP_REGIONS		15
+#define MAX_MMAP_REGIONS		17
 #define MAX_XLAT_TABLES			(MAX_MMAP_REGIONS + BL31SRAM_MAX_PAGES)
 #endif
 #if defined IMAGE_BL33
diff --git a/plat/nxp/s32g/s32g_bl31.c b/plat/nxp/s32g/s32g_bl31.c
index f98f414ac..9561f1c5a 100644
--- a/plat/nxp/s32g/s32g_bl31.c
+++ b/plat/nxp/s32g/s32g_bl31.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2020 NXP
+ * Copyright 2019-2021 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -68,10 +68,14 @@ static const mmap_region_t s32g_mmap[] = {
 			MT_DEVICE | MT_RW),
 	MAP_REGION_FLAT(SSRAMC_BASE_ADDR, SRAMC_SIZE,
 			MT_DEVICE | MT_RW),
-	/* This will be replaced by BL31SRAM */
+	MAP_REGION2(BL31SRAM_BASE, BL31SRAM_BASE,
+		    MMU_ROUND_UP_TO_4K(BL31SRAM_SIZE),
+		    MT_MEMORY | MT_RW, PAGE_SIZE),
 	MAP_REGION2(S32G_BL33_IMAGE_BASE, S32G_BL33_IMAGE_BASE,
 			MMU_ROUND_UP_TO_4K(S32G_BL33_IMAGE_SIZE),
 			MT_MEMORY | MT_RW, PAGE_SIZE),
+	MAP_REGION2(DTB_BASE, DTB_BASE, MMU_ROUND_UP_TO_4K(DTB_SIZE),
+			MT_MEMORY | MT_RW, PAGE_SIZE),
 	MAP_REGION_FLAT(S32G_PMEM_START, S32G_PMEM_LEN,
 			MT_MEMORY | MT_RW | MT_SECURE),
 	MAP_REGION_FLAT(S32G_SCMI_SHARED_MEM, S32G_SCMI_SHARED_MEM_SIZE,
-- 
2.17.1

